crates/edr_provider/src/data.rs

```diff
@@ -1922,6 +1914,141 @@ impl<LoggerErrorT: Debug, TimerT: Clone + TimeSinceEpoch> ProviderData<LoggerErr
+                                        rpc_client,
+                                    )
+                                })
+                                // Go from `Result<Option>` to `Option<Result>` as we're in a flat
+                                // map for an `Option`
+                                .transpose()
+                        })
+                        // Go back from `Option<Result>` to `Result<Option>` which is the expected
+                        // return type in the `block_in_place` to be able to short-circuit.
+                        .transpose()
+                })
+                // Go from `Result<Option>` to `Option<Result>` as we're in a flat
+                // map for an Option
+                .transpose()
+            })
+            // We have an `Option<Result>`, default with another `Option<Result>` from the generic
+            // blockchain interface if it's `None`.
+            .or_else(|| {
+                self.blockchain
+                    // Returns `Result<Option>`
+                    .block_by_transaction_hash(transaction_hash)
+                    .map_err(ProviderError::<LoggerErrorT>::Blockchain)
+                    // We need to return an `Option<Result>`
+                    .transpose()
+                    .map(|block| {
+                        // Map the value in the result then pass the result back out so the error
+                        // can be propagated if any.
+                        block.map(|block| {
+                            let transactions = block.transactions().to_vec();
+                            let header = block.header().clone();
+                            (header, transactions)
+                        })
+                    })
+            })
+            // Go from `Option<Result>` to `Result<Option>` to short-circuit the error.
+            .transpose()?
+            // If we couldn't find the transaction in the remote or local blockchains through the
+            // generic blockchain interface, then it's definitely invalid.
+            .ok_or_else(|| ProviderError::<LoggerErrorT>::InvalidTransactionHash(*transaction_hash))
+    }
+
+    fn block_data_for_debug_trance_transaction_from_block_hash(
+        &self,
+        block_hash: &B256,
+        transaction_hash: &B256,
+        rpc_client: &Arc<EthRpcClient<L1ChainSpec>>,
+    ) -> Result<Option<(Header, Vec<Signed>)>, ProviderError<LoggerErrorT>> {
+        let mut rpc_block = self
+            .runtime_handle
+            .block_on(rpc_client.get_block_by_hash_with_transaction_data(*block_hash))?
+            .ok_or_else(|| {
+                // If the remote returned a transaction for the transaction hash, but the block
+                // is not found, we treat the transaction hash as invalid.
+                ProviderError::<LoggerErrorT>::InvalidTransactionHash(*transaction_hash)
+            })?;
+
+        // We only need the header from the block later, so it's safe to take the
```

Comment:
```suggestion
        // SAFETY: We only need the header from the block later, so it's safe to take the
```

I think it would be good to be explicit about this being related to safety. It makes it easier to search for and validate.

An unobservant editor might change the below code to return more than just the header of the remote block, breaking the "promise" that the `RemoteBlock` matches the received RPC block.

Reply:
Fixed in https://github.com/NomicFoundation/edr/pull/606/commits/0e66f07f3fce6f2646fb479b538d003caf5ab95e